File: src/scripts/lib/data-reader.ts
import JSBI from 'jsbi'

const { BigInt } = JSBI
const big32 = BigInt(32)

const readUInt64LE = (buf: Buffer, byteOffset: number) => {
  const wh = BigInt(buf.readUInt32LE(byteOffset + 4))
  const wl = BigInt(buf.readUInt32LE(byteOffset))
  return JSBI.add(JSBI.leftShift(wh, big32), wl)
}

export class DataReader {
  public pos = 0

  constructor(private buf: Buffer) {}

  skip(length: number) {
    this.pos += length
    return this
  }

  get is_end() {
    return this.pos >= this.buf.length
  }

  u8() {
    const ret = this.buf.readUInt8(this.pos)
    this.pos += 1
    return ret
  }

  u16le() {
    const ret = this.buf.readUInt16LE(this.pos)
    this.pos += 2
    return ret
  }

  u32le() {
    const ret = this.buf.readUInt32LE(this.pos)
    this.pos += 4
    return ret
  }

  u64le() {
    const ret = readUInt64LE(this.buf, this.pos)
    this.pos += 8
    return ret
  }

  i8() {
    const ret = this.buf.readInt8(this.pos)
    this.pos += 1
    return ret
  }

  i16le() {
    const ret = this.buf.readInt16LE(this.pos)
    this.pos += 2
    return ret
  }

  i32le() {
    const ret = this.buf.readInt32LE(this.pos)
    this.pos += 4
    return ret
  }

  f32le() {
    const ret = this.buf.readFloatLE(this.pos)
    this.pos += 4
    return ret
  }

  f64le() {
    const ret = this.buf.readDoubleLE(this.pos)
    this.pos += 8
    return ret
  }

  lstring(length?: number) {
    if (length == null) {
      length = this.buf.length - this.pos
    }

    const ret = this.buf.toString('binary', this.pos, this.pos + length)
    this.pos += length
    return ret
  }

  zstring(terminator = 0) {
    const idx = Array.prototype.indexOf.call(this.buf, terminator, this.pos)

    if (idx == -1) {
      return this.lstring()
    }

    const str = this.lstring(idx - this.pos)
    this.pos++
    return str
  }

  data() {
    const ret = this.buf.slice(this.pos)
    this.pos += ret.byteLength
    return ret
  }
}

File: src/protocols/lib/DataReader.lua
local Object = require("lib/classic")

local DataReader = Object:extend()

function DataReader:new(buf)
  self.buf = buf
  self.pos = 1
end

function DataReader:is_end()
  return self.pos > #self.buf
end

function DataReader:skip(length)
  self.pos = self.pos + length
end

function DataReader:u8()
  local val, pos = string.unpack("I1", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:u16le()
  local val, pos = string.unpack("<I2", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:u32le()
  local val, pos = string.unpack("<I4", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:u64le()
  local val, pos = string.unpack("<I8", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:i8()
  local val, pos = string.unpack("i1", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:i16le()
  local val, pos = string.unpack("<i2", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:i32le()
  local val, pos = string.unpack("<i4", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:i64le()
  local val, pos = string.unpack("<i8", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:f32le()
  local val, pos = string.unpack("<f", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:f64le()
  local val, pos = string.unpack("<d", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:string(length)
  local val, pos = string.unpack("c" .. length, self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:lstring()
  local val, pos = string.unpack("s1", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:zstring()
  local val, pos = string.unpack("z", self.buf, self.pos)
  self.pos = pos
  return val
end

function DataReader:data()
  local data = self.buf:sub(self.pos)
  self.pos = #self.buf
  return data
end

return DataReader
